File: version/src/test/java/com/cedarsoft/VersionRangeTest.java
/**
 * Copyright (C) cedarsoft GmbH.
 *
 * Licensed under the GNU General Public License version 3 (the "License")
 * with Classpath Exception; you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *         http://www.cedarsoft.org/gpl3ce
 *         (GPL 3 with Classpath Exception)
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 only, as
 * published by the Free Software Foundation. cedarsoft GmbH designates this
 * particular file as subject to the "Classpath" exception as provided
 * by cedarsoft GmbH in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 3 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 3 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact cedarsoft GmbH, 72810 Gomaringen, Germany,
 * or visit www.cedarsoft.com if you need additional information or
 * have any questions.
 */

package com.cedarsoft;

import org.testng.annotations.*;

import static org.testng.Assert.*;

/**
 *
 */
public class VersionRangeTest {
  @Test
  public void testConstructor() {
    VersionRange.from( 1, 0, 0 ).to( 2, 0, 0 );
    VersionRange.from( 1, 0, 0 ).to( 1, 0, 0 );

    try {
      VersionRange.from( 1, 0, 1 ).to( 1, 0, 0 );
      fail( "Where is the Exception" );
    } catch ( Exception ignore ) {
    }
  }

  @Test
  public void testFluent() {
    assertEquals( VersionRange.from( 1, 0, 0 ).to( 2, 0, 0 ), new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ) ) );
  }

  @Test
  public void testMinMax() {
    {
      VersionRange range = new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ) );
      assertEquals( range.getMin(), new Version( 1, 0, 0 ) );
      assertEquals( range.getMax(), new Version( 2, 0, 0 ) );
    }
    {
      VersionRange range = new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), false, false );
      assertEquals( range.getMin(), new Version( 1, 0, 0 ) );
      assertEquals( range.getMax(), new Version( 2, 0, 0 ) );
    }
  }

  @Test
  public void testOverlap() {
    VersionRange range = new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ) );

    assertTrue( range.overlaps( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ) ) ) );
    assertTrue( range.overlaps( new VersionRange( new Version( 1, 0, 0 ), new Version( 1, 0, 0 ) ) ) );
    assertTrue( range.overlaps( new VersionRange( new Version( 2, 0, 0 ), new Version( 2, 0, 0 ) ) ) );

    assertFalse( range.overlaps( new VersionRange( new Version( 0, 0, 0 ), new Version( 0, 99, 99 ) ) ) );
    assertFalse( range.overlaps( new VersionRange( new Version( 0, 0, 0 ), new Version( 1, 0, 0 ), true, false ) ) );
    assertTrue( range.overlaps( new VersionRange( new Version( 0, 0, 0 ), new Version( 1, 0, 0 ), true, true ) ) );

    assertFalse( range.overlaps( new VersionRange( new Version( 2, 0, 1 ), new Version( 3, 0, 0 ) ) ) );
    assertFalse( range.overlaps( new VersionRange( new Version( 2, 0, 0 ), new Version( 3, 0, 0 ), false, true ) ) );
    assertTrue( range.overlaps( new VersionRange( new Version( 2, 0, 0 ), new Version( 3, 0, 0 ), true, true ) ) );
  }

  @Test
  public void testContains() {
    assertTrue( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ) ).containsCompletely( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ) ) ) );
    assertTrue( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ) ).containsCompletely( new VersionRange( new Version( 1, 0, 0 ), new Version( 1, 0, 0 ) ) ) );
    assertTrue( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ) ).containsCompletely( new VersionRange( new Version( 2, 0, 0 ), new Version( 2, 0, 0 ) ) ) );

    assertFalse( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), false, true ).containsCompletely( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ) ) ) );
    assertFalse( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, false ).containsCompletely( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ) ) ) );

    assertTrue( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, true ).containsCompletely( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), false, false ) ) );
    assertTrue( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, true ).containsCompletely( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, false ) ) );
    assertTrue( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, true ).containsCompletely( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), false, true ) ) );
    assertTrue( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, true ).containsCompletely( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, true ) ) );

    assertFalse( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), false, true ).containsCompletely( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, true ) ) );
    assertFalse( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), false, true ).containsCompletely( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, false ) ) );
    assertFalse( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, false ).containsCompletely( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, true ) ) );
    assertFalse( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, false ).containsCompletely( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), false, true ) ) );


    assertFalse( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), false, false ).containsCompletely( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, true ) ) );
  }

  @Test
  public void testExclude() {
    {
      VersionRange range = new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, true );
      assertTrue( range.contains( new Version( 1, 0, 0 ) ) );
      assertTrue( range.contains( new Version( 1, 0, 1 ) ) );
      assertTrue( range.contains( new Version( 1, 99, 99 ) ) );
      assertTrue( range.contains( new Version( 2, 0, 0 ) ) );
      assertEquals( range.toString(), "[1.0.0-2.0.0]" );
    }

    {
      VersionRange range = new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), true, false );
      assertTrue( range.contains( new Version( 1, 0, 0 ) ) );
      assertTrue( range.contains( new Version( 1, 0, 1 ) ) );
      assertTrue( range.contains( new Version( 1, 99, 99 ) ) );
      assertFalse( range.contains( new Version( 2, 0, 0 ) ) );
      assertEquals( range.toString(), "[1.0.0-2.0.0[" );
    }

    {
      VersionRange range = new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), false, true );
      assertFalse( range.contains( new Version( 1, 0, 0 ) ) );
      assertTrue( range.contains( new Version( 1, 0, 1 ) ) );
      assertTrue( range.contains( new Version( 1, 99, 99 ) ) );
      assertTrue( range.contains( new Version( 2, 0, 0 ) ) );
      assertEquals( range.toString(), "]1.0.0-2.0.0]" );
    }

    {
      VersionRange range = new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 0, 0 ), false, false );
      assertFalse( range.contains( new Version( 1, 0, 0 ) ) );
      assertTrue( range.contains( new Version( 1, 0, 1 ) ) );
      assertTrue( range.contains( new Version( 1, 99, 99 ) ) );
      assertFalse( range.contains( new Version( 2, 0, 0 ) ) );
      assertEquals( range.toString(), "]1.0.0-2.0.0[" );
    }
  }

  @Test
  public void testIt() {
    VersionRange range = new VersionRange( new Version( 1, 0, 0 ), new Version( 1, 1, 90 ) );

    assertTrue( range.contains( new Version( 1, 0, 0 ) ) );
    assertTrue( range.contains( new Version( 1, 1, 0 ) ) );
    assertTrue( range.contains( new Version( 1, 1, 89 ) ) );
    assertTrue( range.contains( new Version( 1, 1, 90 ) ) );

    assertFalse( range.contains( new Version( 1, 1, 91 ) ) );
    assertFalse( range.contains( new Version( 1, 2, 0 ) ) );
    assertFalse( range.contains( new Version( 0, 99, 99 ) ) );
  }

  @Test
  public void testEquals() {
    assertEquals( new VersionRange( new Version( 1, 0, 0 ), new Version( 1, 1, 90 ) ), new VersionRange( new Version( 1, 0, 0 ), new Version( 1, 1, 90 ) ) );
    assertEquals( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 1, 90 ) ), new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 1, 90 ) ) );

    assertFalse( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 1, 90 ) ).equals( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 1, 91 ) ) ) );
    assertFalse( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 1, 90 ) ).equals( new VersionRange( new Version( 1, 0, 0 ), new Version( 1, 1, 90 ) ) ) );
    assertFalse( new VersionRange( new Version( 1, 0, 1 ), new Version( 2, 1, 90 ) ).equals( new VersionRange( new Version( 1, 0, 0 ), new Version( 2, 1, 90 ) ) ) );
  }

  @Test
  public void testToString() {
    assertEquals( new VersionRange( new Version( 1, 0, 0 ), new Version( 1, 1, 90 ) ).toString(), "[1.0.0-1.1.90]" );
  }
}

File: commons/version/src/jvmTest/kotlin/com/cedarsoft/version/VersionRangeTest.kt
package com.cedarsoft.version

import com.cedarsoft.version.Version.Companion.valueOf
import com.cedarsoft.version.VersionRange.Companion.from
import com.cedarsoft.version.VersionRange.Companion.fromVersions
import com.cedarsoft.version.VersionRange.Companion.single
import com.google.common.collect.ImmutableList
import org.assertj.core.api.Assertions
import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.lang.Exception
import java.lang.IllegalArgumentException

/**
 *
 */
class VersionRangeTest {
  @Test
  fun testSame() {
    val (min, max) = from(1, 1, 1).to(1, 1, 1)
    Assertions.assertThat(min).isEqualTo(valueOf(1, 1, 1))
    Assertions.assertThat(max).isEqualTo(valueOf(1, 1, 1))
  }

  @Test
  @Throws(Exception::class)
  fun fromList() {
    Assertions.assertThat(fromVersions(ImmutableList.of(valueOf(1, 0, 0), valueOf(2, 0, 0)))).isEqualTo(from(1, 0, 0).to(2, 0, 0))
    try {
      fromVersions(ImmutableList.of())
      fail<Any>("Where is the Exception")
    } catch (e: IllegalArgumentException) {
    }
    Assertions.assertThat(fromVersions(ImmutableList.of(valueOf(1, 0, 0)))).isEqualTo(from(1, 0, 0).to(1, 0, 0))
    Assertions.assertThat(fromVersions(ImmutableList.of(valueOf(2, 0, 0), valueOf(1, 0, 0)))).isEqualTo(from(1, 0, 0).to(2, 0, 0))
    Assertions.assertThat(fromVersions(ImmutableList.of(valueOf(1, 5, 0), valueOf(2, 0, 0), valueOf(1, 0, 0)))).isEqualTo(from(1, 0, 0).to(2, 0, 0))
  }

  @Test
  fun testConstructor() {
    from(1, 0, 0).to(2, 0, 0)
    from(1, 0, 0).to(1, 0, 0)
    try {
      from(1, 0, 1).to(1, 0, 0)
      fail<Any>("Where is the Exception")
    } catch (ignore: Exception) {
    }
  }

  @Test
  fun testFormat() {
    MatcherAssert.assertThat(single(1, 0, 0).format(), CoreMatchers.`is`("[1.0.0]"))
    MatcherAssert.assertThat(from(1, 0, 0).to(1, 0, 0).format(), CoreMatchers.`is`("[1.0.0]"))
    MatcherAssert.assertThat(from(1, 0, 0).to(2, 0, 0).format(), CoreMatchers.`is`("[1.0.0-2.0.0]"))
  }

  @Test
  fun testSingle() {
    MatcherAssert.assertThat(single(1, 0, 0).max, CoreMatchers.`is`(valueOf(1, 0, 0)))
    MatcherAssert.assertThat(single(1, 1, 0).max, CoreMatchers.`is`(valueOf(1, 1, 0)))
    MatcherAssert.assertThat(single(valueOf(1, 0, 0)).max, CoreMatchers.`is`(valueOf(1, 0, 0)))
    MatcherAssert.assertThat(single(valueOf(1, 1, 0)).max, CoreMatchers.`is`(valueOf(1, 1, 0)))
  }

  @Test
  fun testFluent() {
    assertEquals(from(1, 0, 0).to(2, 0, 0), VersionRange(Version(1, 0, 0), Version(2, 0, 0)))
  }

  @Test
  fun testMinMax() {
    run {
      val (min, max) = VersionRange(Version(1, 0, 0), Version(2, 0, 0))
      assertEquals(min, Version(1, 0, 0))
      assertEquals(max, Version(2, 0, 0))
    }
    run {
      val (min, max) = VersionRange(Version(1, 0, 0), Version(2, 0, 0), false, false)
      assertEquals(min, Version(1, 0, 0))
      assertEquals(max, Version(2, 0, 0))
    }
  }

  @Test
  fun testOverlap() {
    val range = VersionRange(Version(1, 0, 0), Version(2, 0, 0))
    assertTrue(range.overlaps(VersionRange(Version(1, 0, 0), Version(2, 0, 0))))
    assertTrue(range.overlaps(VersionRange(Version(1, 0, 0), Version(1, 0, 0))))
    assertTrue(range.overlaps(VersionRange(Version(2, 0, 0), Version(2, 0, 0))))
    assertFalse(range.overlaps(VersionRange(Version(0, 0, 0), Version(0, 99, 99))))
    assertFalse(range.overlaps(VersionRange(Version(0, 0, 0), Version(1, 0, 0), true, false)))
    assertTrue(range.overlaps(VersionRange(Version(0, 0, 0), Version(1, 0, 0), true, true)))
    assertFalse(range.overlaps(VersionRange(Version(2, 0, 1), Version(3, 0, 0))))
    assertFalse(range.overlaps(VersionRange(Version(2, 0, 0), Version(3, 0, 0), false, true)))
    assertTrue(range.overlaps(VersionRange(Version(2, 0, 0), Version(3, 0, 0), true, true)))
  }

  @Test
  fun testContains() {
    assertTrue(VersionRange(Version(1, 0, 0), Version(2, 0, 0)).containsCompletely(VersionRange(Version(1, 0, 0), Version(2, 0, 0))))
    assertTrue(VersionRange(Version(1, 0, 0), Version(2, 0, 0)).containsCompletely(VersionRange(Version(1, 0, 0), Version(1, 0, 0))))
    assertTrue(VersionRange(Version(1, 0, 0), Version(2, 0, 0)).containsCompletely(VersionRange(Version(2, 0, 0), Version(2, 0, 0))))
    assertFalse(VersionRange(Version(1, 0, 0), Version(2, 0, 0), false, true).containsCompletely(VersionRange(Version(1, 0, 0), Version(2, 0, 0))))
    assertFalse(VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, false).containsCompletely(VersionRange(Version(1, 0, 0), Version(2, 0, 0))))
    assertTrue(VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, true).containsCompletely(VersionRange(Version(1, 0, 0), Version(2, 0, 0), false, false)))
    assertTrue(VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, true).containsCompletely(VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, false)))
    assertTrue(VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, true).containsCompletely(VersionRange(Version(1, 0, 0), Version(2, 0, 0), false, true)))
    assertTrue(VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, true).containsCompletely(VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, true)))
    assertFalse(VersionRange(Version(1, 0, 0), Version(2, 0, 0), false, true).containsCompletely(VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, true)))
    assertFalse(VersionRange(Version(1, 0, 0), Version(2, 0, 0), false, true).containsCompletely(VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, false)))
    assertFalse(VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, false).containsCompletely(VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, true)))
    assertFalse(VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, false).containsCompletely(VersionRange(Version(1, 0, 0), Version(2, 0, 0), false, true)))
    assertFalse(VersionRange(Version(1, 0, 0), Version(2, 0, 0), false, false).containsCompletely(VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, true)))
  }

  @Test
  fun testExclude() {
    run {
      val range = VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, true)
      assertTrue(range.contains(Version(1, 0, 0)))
      assertTrue(range.contains(Version(1, 0, 1)))
      assertTrue(range.contains(Version(1, 99, 99)))
      assertTrue(range.contains(Version(2, 0, 0)))
      assertEquals("[1.0.0-2.0.0]", range.toString())
    }
    run {
      val range = VersionRange(Version(1, 0, 0), Version(2, 0, 0), true, false)
      assertTrue(range.contains(Version(1, 0, 0)))
      assertTrue(range.contains(Version(1, 0, 1)))
      assertTrue(range.contains(Version(1, 99, 99)))
      assertFalse(range.contains(Version(2, 0, 0)))
      assertEquals("[1.0.0-2.0.0[", range.toString())
    }
    run {
      val range = VersionRange(Version(1, 0, 0), Version(2, 0, 0), false, true)
      assertFalse(range.contains(Version(1, 0, 0)))
      assertTrue(range.contains(Version(1, 0, 1)))
      assertTrue(range.contains(Version(1, 99, 99)))
      assertTrue(range.contains(Version(2, 0, 0)))
      assertEquals("]1.0.0-2.0.0]", range.toString())
    }
    run {
      val range = VersionRange(Version(1, 0, 0), Version(2, 0, 0), false, false)
      assertFalse(range.contains(Version(1, 0, 0)))
      assertTrue(range.contains(Version(1, 0, 1)))
      assertTrue(range.contains(Version(1, 99, 99)))
      assertFalse(range.contains(Version(2, 0, 0)))
      assertEquals("]1.0.0-2.0.0[", range.toString())
    }
  }

  @Test
  fun testIt() {
    val range = VersionRange(Version(1, 0, 0), Version(1, 1, 90))
    assertTrue(range.contains(Version(1, 0, 0)))
    assertTrue(range.contains(Version(1, 1, 0)))
    assertTrue(range.contains(Version(1, 1, 89)))
    assertTrue(range.contains(Version(1, 1, 90)))
    assertFalse(range.contains(Version(1, 1, 91)))
    assertFalse(range.contains(Version(1, 2, 0)))
    assertFalse(range.contains(Version(0, 99, 99)))
  }

  @Test
  fun testEquals() {
    assertEquals(VersionRange(Version(1, 0, 0), Version(1, 1, 90)), VersionRange(Version(1, 0, 0), Version(1, 1, 90)))
    assertEquals(VersionRange(Version(1, 0, 0), Version(2, 1, 90)), VersionRange(Version(1, 0, 0), Version(2, 1, 90)))
    assertNotEquals(VersionRange(Version(1, 0, 0), Version(2, 1, 90)), VersionRange(Version(1, 0, 0), Version(2, 1, 91)))
    assertNotEquals(VersionRange(Version(1, 0, 0), Version(2, 1, 90)), VersionRange(Version(1, 0, 0), Version(1, 1, 90)))
    assertNotEquals(VersionRange(Version(1, 0, 1), Version(2, 1, 90)), VersionRange(Version(1, 0, 0), Version(2, 1, 90)))
  }

  @Test
  fun testToString() {
    assertEquals("[1.0.0-1.1.90]", VersionRange(Version(1, 0, 0), Version(1, 1, 90)).toString())
  }

}
